src/DateObject.ts

```diff
@@ -0,0 +1,351 @@
+				},
+
+				get milliseconds(): number {
+					return self._date.getUTCMilliseconds();
+				},
+				set milliseconds(milliseconds: number) {
+					self._date.setUTCMilliseconds(milliseconds);
+				},
+
+				get dayOfWeek(): number {
+					return self._date.getUTCDay();
+				},
+
+				toString: function (): string {
+					return self._date.toUTCString();
+				}
+			},
+			enumerable: true
+		});
+	}
+
+	get isLeapYear(): boolean {
+		return isLeapYear(this.year);
+	}
+
+	get daysInMonth(): number {
+		var month = this.month;
+
+		if (month === 2 && this.isLeapYear) {
+			return 29;
+		}
+		return days[month];
+	}
+
+	get year(): number {
+		return this._date.getFullYear();
+	}
+	set year(year: number) {
+		var dayOfMonth = this.dayOfMonth;
+
+		this._date.setFullYear(year);
+
+		if (this.dayOfMonth < dayOfMonth) {
+			this.dayOfMonth = 0;
+		}
+	}
+
+	get month(): number {
+		return this._date.getMonth() + 1;
+	}
+	set month(month: number) {
+		var dayOfMonth = this.dayOfMonth;
+
+		this._date.setMonth(month - 1);
+
+		if (this.dayOfMonth < dayOfMonth) {
+			this.dayOfMonth = 0;
+		}
+	}
+
+	get dayOfMonth(): number {
+		return this._date.getDate();
+	}
+	set dayOfMonth(day: number) {
+		this._date.setDate(day);
+	}
+
+	get hours(): number {
+		return this._date.getHours();
+	}
+	set hours(hours: number) {
+		this._date.setHours(hours);
+	}
+
+	get minutes(): number {
+		return this._date.getMinutes();
+	}
+	set minutes(minutes: number) {
+		this._date.setMinutes(minutes);
+	}
+
+	get seconds(): number {
+		return this._date.getSeconds();
+	}
+	set seconds(seconds: number) {
+		this._date.setSeconds(seconds);
+	}
+
+	get milliseconds(): number {
+		return this._date.getMilliseconds();
+	}
+	set milliseconds(milliseconds: number) {
+		this._date.setMilliseconds(milliseconds);
+	}
+
+	get time(): number {
+		return this._date.getTime();
+	}
+	set time(time: number) {
+		this._date.setTime(time);
+	}
+
+	get dayOfWeek(): number {
+		return this._date.getDay();
+	}
+	get timezoneOffset(): number {
+		return this._date.getTimezoneOffset();
+	}
+
+	add(value: number): DateObject;
+	add(value: Duration): DateObject;
+	add(value: DateLike): DateObject;
+	add(value: any): DateObject {
+		var result = new DateObject(this);
+
+		if (value instanceof Duration) {
+			result.time += (<Duration> value).time;
+		}
+		else if (typeof value === 'object') {
+			// Anything date-like may have year and month values that need to be handled
+			// as dates rather than durations to avoid overshoots
+			operationOrder.forEach((property: string): void => {
+				if (!(property in value)) {
+					return;
+				}
+
+				(<any> result)[property] += (<any> value)[property];
+			});
+		}
+		else if (typeof value === 'number') {
+			result.time += (<number> value);
+		}
+		else {
+			// Attempt to put value into a DateObject and translate to a time
+			result.time += new DateObject((<any> value)).time;
+		}
+
+		return result;
+	}
+
+	difference(value: number): Duration;
```

Comment:
Consider renaming difference() to until() to match with Duration.until() and disambiguate from subtract(). difference() is ambiguous in general practice when returning a Duration once the difference goes beyond the border of a month or leap year. until() is more direct in answering date1.until(date2) -> duration.